File: research/cloud_10coin_backtest/g9/data_contract.py
from __future__ import annotations

from datetime import datetime
import hashlib
import json
from typing import Any, Mapping


CONTRACT_VERSION = 1
ALLOWED_FRESHNESS = {"FRESH", "DEGRADED", "STALE", "UNKNOWN"}
AUTHORITY_SCOPE = "research_evidence"


def _iso(value: datetime | str) -> str:
    if isinstance(value, str):
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
        if parsed.tzinfo is None:
            raise ValueError("timestamp must be timezone-aware")
        return parsed.isoformat()
    if value.tzinfo is None:
        raise ValueError("timestamp must be timezone-aware")
    return value.isoformat()


def _parse_time(value: Any) -> datetime | None:
    if not isinstance(value, str) or not value:
        return None
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError:
        return None
    if parsed.tzinfo is None:
        return None
    return parsed


def _canonical(payload: Any) -> str:
    return json.dumps(payload, sort_keys=True, separators=(",", ":"), ensure_ascii=True, allow_nan=False)


def compute_payload_hash(payload: Any) -> str:
    return hashlib.sha256(_canonical(payload).encode("utf-8")).hexdigest()


def build_envelope(
    *,
    kind: str,
    source: str,
    source_sha: str,
    event_time: datetime | str,
    ingest_time: datetime | str,
    freshness: str,
    payload: Mapping[str, Any] | list[Any],
    provenance: Mapping[str, Any],
) -> dict[str, Any]:
    normalized_freshness = str(freshness).upper()
    envelope = {
        "contract_version": CONTRACT_VERSION,
        "kind": str(kind),
        "source": str(source),
        "source_sha": str(source_sha),
        "event_time": _iso(event_time),
        "ingest_time": _iso(ingest_time),
        "freshness": normalized_freshness,
        "authority": {
            "scope": AUTHORITY_SCOPE,
            "execution": "none",
        },
        "research_only": True,
        "production_execution_authority": False,
        "provenance": dict(provenance),
        "payload": payload,
    }
    envelope["payload_hash"] = compute_payload_hash(payload)
    return envelope


def validate_envelope(payload: Mapping[str, Any]) -> list[str]:
    if not isinstance(payload, Mapping):
        return ["ENVELOPE_INVALID"]
    errors: list[str] = []

    if payload.get("contract_version") != CONTRACT_VERSION:
        errors.append("CONTRACT_VERSION_INVALID")
    if not isinstance(payload.get("kind"), str) or not payload.get("kind"):
        errors.append("KIND_REQUIRED")
    if not isinstance(payload.get("source"), str) or not payload.get("source"):
        errors.append("SOURCE_REQUIRED")
    if not isinstance(payload.get("source_sha"), str) or not payload.get("source_sha"):
        errors.append("SOURCE_SHA_REQUIRED")
    if str(payload.get("freshness") or "").upper() not in ALLOWED_FRESHNESS:
        errors.append("FRESHNESS_INVALID")

    event_time = _parse_time(payload.get("event_time"))
    ingest_time = _parse_time(payload.get("ingest_time"))
    if event_time is None:
        errors.append("EVENT_TIME_INVALID")
    if ingest_time is None:
        errors.append("INGEST_TIME_INVALID")
    if event_time is not None and ingest_time is not None and event_time > ingest_time:
        errors.append("EVENT_AFTER_INGEST")

    authority = payload.get("authority")
    if not isinstance(authority, Mapping):
        errors.append("AUTHORITY_INVALID")
    else:
        if authority.get("scope") != AUTHORITY_SCOPE:
            errors.append("AUTHORITY_SCOPE_INVALID")
        if authority.get("execution") != "none":
            errors.append("EXECUTION_AUTHORITY_FORBIDDEN")
    if payload.get("research_only") is not True:
        errors.append("RESEARCH_ONLY_REQUIRED")
    if payload.get("production_execution_authority") is not False:
        errors.append("PRODUCTION_EXECUTION_AUTHORITY_FORBIDDEN")
    if not isinstance(payload.get("provenance"), Mapping):
        errors.append("PROVENANCE_INVALID")

    body = payload.get("payload")
    expected = payload.get("payload_hash")
    if not isinstance(expected, str) or len(expected) != 64:
        errors.append("PAYLOAD_HASH_INVALID")
    else:
        try:
            actual = compute_payload_hash(body)
        except (TypeError, ValueError):
            errors.append("PAYLOAD_INVALID")
        else:
            if actual != expected:
                errors.append("PAYLOAD_HASH_MISMATCH")
    return errors

File: crypto-research-gateway/src/normalization/data-contract.ts
import { createHash } from 'node:crypto';

export const DATA_CONTRACT_VERSION = 1 as const;
export const DATA_AUTHORITY_SCOPE = 'research_evidence' as const;
export type DataFreshness = 'FRESH' | 'DEGRADED' | 'STALE' | 'UNKNOWN';

export type DataEnvelope = {
  contract_version: 1;
  kind: string;
  source: string;
  source_sha: string;
  event_time: string;
  ingest_time: string;
  freshness: DataFreshness;
  authority: {
    scope: typeof DATA_AUTHORITY_SCOPE;
    execution: 'none';
  };
  research_only: true;
  production_execution_authority: false;
  provenance: Record<string, unknown>;
  payload: unknown;
  payload_hash: string;
};

function normalize(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(normalize);
  if (value && typeof value === 'object') {
    return Object.fromEntries(
      Object.entries(value as Record<string, unknown>)
        .sort(([a], [b]) => a.localeCompare(b))
        .map(([key, child]) => [key, normalize(child)]),
    );
  }
  return value;
}

function canonical(value: unknown): string {
  return JSON.stringify(normalize(value));
}

export function computePayloadHash(payload: unknown): string {
  return createHash('sha256').update(canonical(payload), 'utf8').digest('hex');
}

export function buildDataEnvelope(input: {
  kind: string;
  source: string;
  sourceSha: string;
  eventTime: string;
  ingestTime: string;
  freshness: DataFreshness;
  payload: unknown;
  provenance: Record<string, unknown>;
}): DataEnvelope {
  return {
    contract_version: DATA_CONTRACT_VERSION,
    kind: input.kind,
    source: input.source,
    source_sha: input.sourceSha,
    event_time: input.eventTime,
    ingest_time: input.ingestTime,
    freshness: input.freshness,
    authority: { scope: DATA_AUTHORITY_SCOPE, execution: 'none' },
    research_only: true,
    production_execution_authority: false,
    provenance: input.provenance,
    payload: input.payload,
    payload_hash: computePayloadHash(input.payload),
  };
}

export function validateDataEnvelope(payload: unknown): string[] {
  if (!payload || typeof payload !== 'object' || Array.isArray(payload)) return ['ENVELOPE_INVALID'];
  const row = payload as Record<string, unknown>;
  const errors: string[] = [];
  if (row.contract_version !== DATA_CONTRACT_VERSION) errors.push('CONTRACT_VERSION_INVALID');
  if (typeof row.kind !== 'string' || !row.kind) errors.push('KIND_REQUIRED');
  if (typeof row.source !== 'string' || !row.source) errors.push('SOURCE_REQUIRED');
  if (typeof row.source_sha !== 'string' || !row.source_sha) errors.push('SOURCE_SHA_REQUIRED');
  if (!['FRESH', 'DEGRADED', 'STALE', 'UNKNOWN'].includes(String(row.freshness))) errors.push('FRESHNESS_INVALID');

  const eventTime = typeof row.event_time === 'string' ? Date.parse(row.event_time) : Number.NaN;
  const ingestTime = typeof row.ingest_time === 'string' ? Date.parse(row.ingest_time) : Number.NaN;
  if (!Number.isFinite(eventTime)) errors.push('EVENT_TIME_INVALID');
  if (!Number.isFinite(ingestTime)) errors.push('INGEST_TIME_INVALID');
  if (Number.isFinite(eventTime) && Number.isFinite(ingestTime) && eventTime > ingestTime) errors.push('EVENT_AFTER_INGEST');

  const authority = row.authority;
  if (!authority || typeof authority !== 'object' || Array.isArray(authority)) {
    errors.push('AUTHORITY_INVALID');
  } else {
    const auth = authority as Record<string, unknown>;
    if (auth.scope !== DATA_AUTHORITY_SCOPE) errors.push('AUTHORITY_SCOPE_INVALID');
    if (auth.execution !== 'none') errors.push('EXECUTION_AUTHORITY_FORBIDDEN');
  }
  if (row.research_only !== true) errors.push('RESEARCH_ONLY_REQUIRED');
  if (row.production_execution_authority !== false) errors.push('PRODUCTION_EXECUTION_AUTHORITY_FORBIDDEN');
  if (!row.provenance || typeof row.provenance !== 'object' || Array.isArray(row.provenance)) errors.push('PROVENANCE_INVALID');

  if (typeof row.payload_hash !== 'string' || row.payload_hash.length !== 64) {
    errors.push('PAYLOAD_HASH_INVALID');
  } else if (computePayloadHash(row.payload) !== row.payload_hash) {
    errors.push('PAYLOAD_HASH_MISMATCH');
  }
  return errors;
}
